src/reaper/mod.rs

```diff
@@ -0,0 +1,221 @@
+//! The underlying system reaper.
+//!
+//! There are two backends:
+//!
+//! - signal, which waits for SIGCHLD.
+//! - wait, which waits directly on a process handle.
+//!
+//! "wait" is preferred, but is not available on all supported Linuxes. So we
+//! test to see if pidfd is supported first. If it is, we use wait. If not, we use
+//! signal.
+
+#![allow(irrefutable_let_patterns)]
+
+/// Enable the waiting reaper.
+#[cfg(any(target_os = "linux", target_os = "android"))]
+macro_rules! cfg_wait {
+    ($($tt:tt)*) => {$($tt)*};
+}
+
+/// Enable the waiting reaper.
+#[cfg(not(any(target_os = "linux", target_os = "android")))]
+macro_rules! cfg_wait {
+    ($($tt:tt)*) => {};
+}
+
+/// Enable signals.
+macro_rules! cfg_signal {
+    ($($tt:tt)*) => {$($tt)*};
+}
+
+cfg_wait! {
+    mod wait;
+}
+
+cfg_signal! {
+    mod signal;
+}
+
+use std::io;
+use std::sync::Mutex;
+
+/// The underlying system reaper.
+pub(crate) enum Reaper {
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// The reaper based on the wait backend.
+    Wait(wait::Reaper),
+
+    /// The reaper based on the signal backend.
+    Signal(signal::Reaper),
+}
+
+/// The wrapper around a child.
+pub(crate) enum ChildGuard {
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// The child guard based on the wait backend.
+    Wait(wait::ChildGuard),
+
+    /// The child guard based on the signal backend.
+    Signal(signal::ChildGuard),
+}
+
+/// A lock on the reaper.
+pub(crate) enum Lock {
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// The wait-based reaper needs no lock.
+    Wait,
+
+    /// The lock for the signal-based reaper.
+    Signal(signal::Lock),
+}
+
+impl Reaper {
+    /// Create a new reaper.
+    pub(crate) fn new() -> Self {
+        cfg_wait! {
+            if wait::available() && !cfg!(async_process_force_signal_backend) {
```

Comment:
wouldn't it make more sense to simply handle `async_process_force_signal_backend` in `cfg_wait!`?

Reply:
For Unixes that don't have wait-based backends, it would be redundant to check this flag. In addition the inner type will already be an `enum` of either `Signal` or `Wait` variants. So matching on this will be redundant.

Reply:
I only intend `async_process_force_signal_backend` to be used for testing of `async_process`. It shouldn't be used in prod builds. I'm fine with not eliminating the dead code here.